lib/flex_algo.h

```diff
@@ -0,0 +1,172 @@
+/*
+ * Flexible Algorithm definitions
+ * Copyright (C) 2022  Hiroki Shirokura, LINE Corporation
+ * Copyright (C) 2022  Masakazu Asama
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; see the file COPYING; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#ifndef _FRR_FLEX_ALGO_H
+#define _FRR_FLEX_ALGO_H
+
+#include "linklist.h"
+#include "prefix.h"
+#include "bitfield.h"
+#include "segment_routing.h"
+
+#define ADMIN_GROUP_PRINT_MAX_SIZE 2048
+#define AFFINITY_NAME_SIZE 256
+#define FLEX_ALGO_PRIO_DEFAULT 128
+#define EXT_ADMIN_GROUP_MAX_POSITIONS 1024
+
+#define CALC_TYPE_SPF 0
+
+/* flex-algo definition flags */
+
+/* M-flag (aka. prefix-metric)
+ * Flex-Algorithm specific prefix and ASBR metric MUST be used
+ */
+#define FAD_FLAG_M 0x80
+
+/*
+ * Metric Type values from draft-ietf-lsr-flex-algo-19
+ */
+enum flex_algo_metric_type {
+	MT_IGP = 0,
+	MT_MIN_UNI_LINK_DELAY = 1,
+	MT_TE_DEFAULT = 2,
+};
+
+struct affinity_map {
+	char name[AFFINITY_NAME_SIZE];
+	uint16_t bit_position;
+};
+
+struct affinity_maps {
+	struct list *maps;
+};
+
+struct admin_group {
+	bitfield_t bitmap;
+};
+
+struct flex_algo {
+	uint8_t algorithm;
+	enum flex_algo_metric_type metric_type;
+	uint8_t calc_type;
+	uint8_t priority;
+	uint8_t flags;
+
+#define FLEX_ALGO_SR_MPLS 0x01
```

Comment:
I don't think these spec-defined values. What's the point of defining in the library header. These prorocol-specific and implementation specific values and should be in ISIS code only.

Reply:
IP, SR-MPLS or SRv6 dataplanes are options for IS-IS and OSPF